Source/Core/Core/HW/HSP/HSP.cpp

```diff
@@ -0,0 +1,63 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/HW/HSP/HSP.h"
+
+#include <memory>
+
+#include "Common/Logging/Log.h"
+#include "Core/Config/MainSettings.h"
+#include "Core/ConfigManager.h"
+#include "Core/HW/HSP/HSP_Device.h"
+
+namespace HSP
+{
+static std::unique_ptr<IHSPDevice> s_device;
+
+void Init()
+{
+  AddDevice(Config::Get(Config::MAIN_HSP_DEVICE));
+}
+
+void Shutdown()
+{
+  RemoveDevice();
+}
+
+u64 Read(u32 address)
+{
+  DEBUG_LOG_FMT(HSP, "HSP read from 0x{:08x}", address);
+  if (s_device)
+    return s_device->Read(address);
+  return 0;
+}
+
+void Write(u32 address, u64 value)
+{
+  DEBUG_LOG_FMT(HSP, "HSP write to 0x{:08x}: 0x{:016x}", address, value);
+  if (s_device)
+    s_device->Write(address, value);
+}
+
+void DoState(PointerWrap& p)
+{
+  if (s_device)
+    s_device->DoState(p);
+}
```

Comment:
I don't think this is going to work properly; if `s_device` is not present, then the save state will have a different format from if it is present (although that probably shouldn't happen, since it should only be possible to load states after `Init()` is called and before `Shutdown()` is called).  More importantly, though, this doesn't handle changing devices if the savestate was saved with a different device selected than when it was loaded.

Here's how the EXI version handles this - it's a bit messy, but to my understanding it solves this issue:

https://github.com/dolphin-emu/dolphin/blob/0f2540a0d1133950467845f20b1e003181147781/Source/Core/Core/HW/EXI/EXI_Channel.cpp#L247-L261

(That code also assumes that device is not null, and in fact EXI only checks it if it's directly using the chip select input since some values of that correspond to no device.  The same logic here could remove the checks on whether `s_device` is present, but I don't feel too strongly about whether that should be done.)

Reply:
This isn't done in exactly the same way, but I think it's fixed now.

Reply:
The new way looks fine to me; EXI's logic is a bit more complicated because it needs to handle hot-swapping (and in some cases making sure games notice that hot-swapping happened by having the device register as via interrupts, but it needs to not do that for a savestate).  I assume it's not normally valid to remove an HSP device while a game is running.

Reply:
Considering you'd need to pick up the GameCube to remove an HSP device, that seems like a safe bet.